### PR TITLE
Reland: Make `torch.empty*` deterministic by filling with NaN or max int

### PR DESCRIPTION
**Pull request by kurtamohler**
Relands #101849 after #104302 reverted it.

torchrec PR https://github.com/pytorch/torchrec/pull/1269 fixes the torchrec failure that caused #101849 to be reverted

Part of #82004

cc @mruberry

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Thanks for pushing the fix on the torchrec side!

**Comment by kurtamohler** [comment]
Hey @albanD, if possible, would you be able to import this PR and run the internal torchrec tests, just to make sure they pass before we merge it?

**Comment by albanD** [comment]
Sure, I'll do a rebase because there was a revert in trunk responsible for the current CI failure.

**Comment by albanD** [comment]
There seems to be other issues where code expects `empty()` to return 0s :(

**Comment by kurtamohler** [comment]
Darn, that's unfortunate. Thanks for the help @albanD, let me know if there's anything I can do

**Comment by albanD** [comment]
First change seems to fix some. Waiting on CI to run again.

**Comment by kurtamohler** [comment]
Awesome, thank you @albanD!

**Pull request closed without merging**
